File: src-tauri/src/models/analysis.rs
// Feature 021: Analysis Data Models
// Defines all data structures for skill analysis results

use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};

/// Main analysis result container
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillAnalysisResult {
    pub analysis_id: String,
    pub skill_name: String,
    pub skill_path: String,
    pub timestamp: DateTime<Utc>,
    pub status: AnalysisStatus,
    pub progress: u8,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub spec_compliance: Option<SpecCompliance>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pda_analysis: Option<PDAAnalysis>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security_review: Option<SecurityReview>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trigger_suggestions: Option<Vec<TriggerSuggestion>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub link_validation: Option<LinkValidation>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AnalysisStatus {
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpecCompliance {
    pub score: u8,
    pub violations: Vec<Violation>,
    pub warnings: Vec<Warning>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Violation {
    pub rule: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fix_suggestion: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Warning {
    pub rule: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommendation: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PDAAnalysis {
    pub score: u8,
    pub token_estimate: usize,
    pub tier_breakdown: TierBreakdown,
    pub recommendations: Vec<String>,
    pub suggested_structure: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TierBreakdown {
    pub metadata_tokens: usize,
    pub orchestrator_tokens: usize,
    pub resource_tokens: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SecurityReview {
    pub score: u8,
    pub unused_permissions: Vec<String>,
    pub high_risk_permissions: Vec<RiskFlag>,
    pub minimum_required: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RiskFlag {
    pub permission: String,
    pub risk_level: RiskLevel,
    pub explanation: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mitigation: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RiskLevel {
    Critical,
    High,
    Medium,
    Low,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TriggerSuggestion {
    pub keyword: String,
    pub relevance_score: u8,
    pub explanation: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinkValidation {
    pub total_links: usize,
    pub valid_links: usize,
    pub broken_links: Vec<BrokenLink>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BrokenLink {
    pub url: String,
    pub line_number: usize,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CLIDetectionResult {
    pub claude_available: bool,
    pub opencode_available: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub claude_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub opencode_path: Option<String>,
}

File: src/types/analysis.ts
// Feature 021: Analysis Type Definitions
// TypeScript interfaces mirroring Rust data models

/**
 * Analysis execution status
 */
export type AnalysisStatus = 'running' | 'completed' | 'failed';

/**
 * Main analysis result container
 */
export interface SkillAnalysisResult {
  /** Unique identifier for this analysis (UUID v4) */
  analysis_id: string;

  /** Skill directory name */
  skill_name: string;

  /** Absolute path to SKILL.md */
  skill_path: string;

  /** Analysis start timestamp (ISO 8601) */
  timestamp: string;

  /** Current analysis status */
  status: AnalysisStatus;

  /** Progress percentage (0-100) */
  progress: number;

  /** Spec compliance results (FR-001) */
  spec_compliance?: SpecCompliance;

  /** PDA analysis results (FR-002, FR-003) */
  pda_analysis?: PDAAnalysis;

  /** Security review results (FR-004) */
  security_review?: SecurityReview;

  /** Trigger keyword suggestions (FR-005) */
  trigger_suggestions?: TriggerSuggestion[];

  /** Link validation results (FR-006) */
  link_validation?: LinkValidation;

  /** Error message if analysis failed */
  error?: string;
}

/**
 * Anthropic Skills Specification compliance results
 */
export interface SpecCompliance {
  /** Overall compliance score (0-100) */
  score: number;

  /** List of specification violations */
  violations: Violation[];

  /** List of warnings */
  warnings: Warning[];
}

/**
 * A specification violation
 */
export interface Violation {
  /** Machine-readable rule identifier */
  rule: string;

  /** Human-readable error message */
  message: string;

  /** Optional fix suggestion */
  fix_suggestion?: string;

  /** Line number where violation occurs */
  line_number?: number;
}

/**
 * A specification warning
 */
export interface Warning {
  /** Machine-readable rule identifier */
  rule: string;

  /** Human-readable warning message */
  message: string;

  /** Optional recommendation */
  recommendation?: string;
}

/**
 * Progressive Disclosure Architecture analysis results
 */
export interface PDAAnalysis {
  /** PDA compliance score (0-100) */
  score: number;

  /** Estimated token count */
  token_estimate: number;

  /** Token breakdown by tier */
  tier_breakdown: TierBreakdown;

  /** Recommendations for improvement */
  recommendations: string[];

  /** Suggested structural changes */
  suggested_structure: string[];

  /** AI-powered insights (only present in LLM-enhanced analysis) */
  ai_insights?: string[];
}

/**
 * Token distribution across PDA tiers
 */
export interface TierBreakdown {
  /** Tokens in frontmatter */
  metadata_tokens: number;

  /** Tokens in main SKILL.md */
  orchestrator_tokens: number;

  /** Tokens in references/ */
  resource_tokens: number;
}

/**
 * Risk severity levels
 */
export type RiskLevel = 'critical' | 'high' | 'medium' | 'low';

/**
 * Permissions security analysis results
 */
export interface SecurityReview {
  /** Security score (0-100) */
  score: number;

  /** Unused permissions */
  unused_permissions: string[];

  /** High-risk permission combinations */
  high_risk_permissions: RiskFlag[];

  /** Minimum required permissions */
  minimum_required: string[];
}

/**
 * A security risk flag
 */
export interface RiskFlag {
  /** Permission name */
  permission: string;

  /** Severity level */
  risk_level: RiskLevel;

  /** Explanation of risk */
  explanation: string;

  /** Optional mitigation advice */
  mitigation?: string;
}

/**
 * A suggested trigger keyword
 */
export interface TriggerSuggestion {
  /** Suggested keyword */
  keyword: string;

  /** Relevance score (0-100) */
  relevance_score: number;

  /** Explanation of relevance */
  explanation: string;
}

/**
 * Link validation results
 */
export interface LinkValidation {
  /** Total links found */
  total_links: number;

  /** Valid (reachable) links */
  valid_links: number;

  /** Broken links */
  broken_links: BrokenLink[];
}

/**
 * A broken link
 */
export interface BrokenLink {
  /** Broken URL or file path */
  url: string;

  /** Line number in SKILL.md */
  line_number: number;

  /** Error message */
  error: string;
}

/**
 * CLI tool detection results
 */
export interface CLIDetectionResult {
  /** Whether Claude CLI is available */
  claude_available: boolean;

  /** Whether OpenCode CLI is available */
  opencode_available: boolean;

  /** Whether Gemini CLI is available */
  gemini_available: boolean;

  /** Path to claude binary */
  claude_path?: string;

  /** Path to opencode binary */
  opencode_path?: string;

  /** Path to gemini binary */
  gemini_path?: string;
}
